internal/project/parsecache.go

```diff
@@ -0,0 +1,99 @@
+package project
+
+import (
+	"sync"
+
+	"github.com/microsoft/typescript-go/internal/ast"
+	"github.com/microsoft/typescript-go/internal/collections"
+	"github.com/microsoft/typescript-go/internal/core"
+	"github.com/microsoft/typescript-go/internal/parser"
+	"github.com/zeebo/xxh3"
+)
+
+type parseCacheKey struct {
+	ast.SourceFileParseOptions
+	scriptKind core.ScriptKind
+}
+
+func newParseCacheKey(
+	options ast.SourceFileParseOptions,
+	scriptKind core.ScriptKind,
+) parseCacheKey {
+	return parseCacheKey{
+		SourceFileParseOptions: options,
+		scriptKind:             scriptKind,
+	}
+}
+
+type parseCacheEntry struct {
+	mu         sync.Mutex
+	sourceFile *ast.SourceFile
+	hash       xxh3.Uint128
+	refCount   int
+}
+
+type ParseCacheOptions struct {
+	// DisableDeletion prevents entries from being removed from the cache.
+	// Used for testing.
+	DisableDeletion bool
+}
+
+type ParseCache struct {
+	Options ParseCacheOptions
+	entries collections.SyncMap[parseCacheKey, *parseCacheEntry]
+}
+
+func (c *ParseCache) Acquire(
+	fh FileContent,
+	opts ast.SourceFileParseOptions,
+	scriptKind core.ScriptKind,
+) *ast.SourceFile {
+	key := newParseCacheKey(opts, scriptKind)
+	entry, loaded := c.loadOrStoreNewLockedEntry(key)
+	defer entry.mu.Unlock()
+	if !loaded || entry.hash != fh.Hash() {
+		// Reparse the file if the hash has changed, or parse for the first time.
+		entry.sourceFile = parser.ParseSourceFile(opts, fh.Content(), scriptKind)
+		entry.hash = fh.Hash()
+	}
+	return entry.sourceFile
+}
+
+func (c *ParseCache) Ref(file *ast.SourceFile) {
```

Comment:
I haven't seen where this is used, but I'm assuming this has ref counting because it's the cache for the actual server, and not the one used for the test harness that should cache foreverish?

Reply:
This replaces DocumentStore, yeah. And I ended up just making an option on it to disable deletion so it can be passed in for tests without implementing an interface.